File: smartgas-debit/backend/internal/services/swit.py
import requests
import json
import sentry_sdk
from app.config import SWIT_BUSINESS, SWIT_API_URL, SWIT_TOKEN, SWIT_API_KEY
from internal.exceptions.swit import SwitNotFundsOrIncorrectException


def get_headers():
    return {
        "business": SWIT_BUSINESS,
        "token": SWIT_TOKEN,
        "x-api-key": SWIT_API_KEY,
    }


def reserve_funds(
    swit_customeer_id, source_id, cvv, last_4, amount, email=None, description=""
):
    headers = get_headers()
    data = {
        "customerId": swit_customeer_id,
        "sourceId": source_id,
        "cvv": cvv,
        "amount": amount,
        "capture": True,
        "cardLastDigits": last_4,
        "description": description,
    }
    response = requests.post(
        f"{SWIT_API_URL}/payments", headers=headers, data=json.dumps(data)
    )

    if response.status_code not in [200, 400]:
        scope = sentry_sdk.get_current_scope()
        scope.set_user({"email": email})
        scope.set_level("error")
        scope.set_transaction_name("SwitPaymentError")
        scope.set_context(
            "Swit",
            {
                "url": SWIT_API_URL,
                "status_code": response.status_code,
            },
        )
        scope.capture_message(
            f"Error while reseerving card funds for user user {email} in swit"
        )
        raise Exception(
            f"Error while reseerving card funds for user user {email} in swit"
        )

    response_data = response.json()

    if response.status_code == 400:
        raise SwitNotFundsOrIncorrectException

    result = response_data["result"]

    return result["transactionId"]


def get_cards_by_user(swit_customeer_id, email=None):
    headers = get_headers()
    response = requests.get(
        f"{SWIT_API_URL}/customers/{swit_customeer_id}/cards", headers=headers
    )

    if response.status_code != 200:
        scope = sentry_sdk.get_current_scope()
        scope.set_user({"email": email})
        scope.set_transaction_name("SwitCardList")
        scope.set_context(
            "Swit",
            {
                "url": SWIT_API_URL,
                "status_code": response.status_code,
            },
        )
        scope.capture_message(
            f"Error while retrieving cards for user user {email} in swit"
        )
        raise Exception(
            f"Unable to get cards for user in swit - {response.status_code}"
        )

    response_data = response.json()
    return response_data["result"]


def create_customer(email, first_name, last_name):
    headers = get_headers()
    data = {
        "email": email,
        "first_name": first_name,
        "last_name": last_name,
    }
    response = requests.post(
        f"{SWIT_API_URL}/customers",
        headers=headers,
        data=json.dumps(data),
    )

    if response.status_code != 200:
        scope = sentry_sdk.get_current_scope()
        scope.set_user(data)
        scope.set_transaction_name("SwitUserCreationError")
        scope.set_context(
            "Swit",
            {
                "url": SWIT_API_URL,
                "status_code": response.status_code,
            },
        )
        scope.capture_message(f"Error while registering user {email} in swit")
        raise Exception(f"Unable to register user in swit - {response.status_code}")

    response_data = response.json()

    if response_data["status"] != "Success":
        scope = sentry_sdk.get_current_scope()
        scope.set_user(data)
        scope.set_transaction_name("SwitUserResponseError")
        scope.set_context(
            "Swit",
            {
                "url": SWIT_API_URL,
                **response_data,
            },
        )
        scope.capture_message(f"Error while registering user {email} in swit")
        raise Exception(f"Unable to register user in swit")

    return response_data["result"]

File: smartgas-payments-composer/backend/internal/services/swit.go
package services

import (
	"bytes"
	"encoding/json"
	"errors"
	"net/http"
	"smartgas-payment/config"
	"smartgas-payment/internal/dto"
	"smartgas-payment/internal/schemas"
	"smartgas-payment/internal/utils"
	"time"

	"github.com/jinzhu/copier"
)

const (
	timeout = 60 // Timeout in seconds for swit
)

var ErrProccesingPayment = errors.New("CVV error or not funds")

type ReserveFundsOpts struct {
	CustomerID  string  `json:"customerId"`
	SourceID    string  `json:"sourceId"`
	Cvv         string  `json:"cvv"`
	Last4       string  `json:"cardLastDigits"`
	Amount      float32 `json:"amount"`
	Description string  `json:"description"`
	Capture     bool    `json:"capture"`
}

//go:generate mockery --name SwitService --filename=mock_swit.go --inpackage=true
type SwitService interface {
	CreateCustomer(*schemas.Customer) (string, error)
	ListCardsByCustomer(string) ([]schemas.SwitSource, error)
	ReserveFunds(ReserveFundsOpts) (string, error)
	CancelFundReservation(string) error
	ConfirmFundReservation(string, float32) error
	DeleteCard(string, string) error
}

type switService struct {
	cfg config.Config
}

func ProvideSwitService(cfg config.Config) *switService {
	return &switService{
		cfg: cfg,
	}
}

func (ss *switService) addHeaders(r *http.Request) {
	r.Header.Add("business", ss.cfg.SwitBusiness)
	r.Header.Add("token", ss.cfg.SwitToken)
	r.Header.Add("x-api-key", ss.cfg.SwitApiKey)
	r.Header.Add("Content-Type", "application/json")
}

func (ss *switService) DeleteCard(cusID string, cardID string) error {
	client := &http.Client{
		Timeout: time.Second * timeout,
	}

	req, err := http.NewRequest(
		"DELETE",
		ss.cfg.SwitBaseUrl+"/customers/"+cusID+"/cards/"+cardID,
		nil,
	)
	if err != nil {
		return err
	}

	ss.addHeaders(req)

	res, err := client.Do(req)
	if err != nil {
		return err
	}

	var data dto.SwitGeneralResponse

	json.NewDecoder(res.Body).Decode(&data)

	if data.Status != "Success" {
		return errors.New("Error deleting card")
	}

	return nil
}

func (ss *switService) ConfirmFundReservation(transID string, amount float32) error {
	client := &http.Client{
		Timeout: time.Second * timeout,
	}

	payload := struct {
		Amount float32 `json:"amount"`
	}{
		Amount: amount,
	}
	payloadData, _ := json.Marshal(payload)

	payloadRequest := bytes.NewReader(payloadData)

	req, err := http.NewRequest("PUT", ss.cfg.SwitBaseUrl+"/payments/"+transID, payloadRequest)
	if err != nil {
		return err
	}

	ss.addHeaders(req)

	res, err := client.Do(req)
	if err != nil {
		return err
	}

	var data dto.SwitGeneralResponse

	json.NewDecoder(res.Body).Decode(&data)

	if data.Status != "Success" {
		return errors.New("Error confirming")
	}

	return nil
}

func (ss *switService) CancelFundReservation(transID string) error {
	client := &http.Client{
		Timeout: time.Second * timeout,
	}

	req, err := http.NewRequest("PUT", ss.cfg.SwitBaseUrl+"/payments/"+transID+"/cancel", nil)
	if err != nil {
		return err
	}

	ss.addHeaders(req)

	res, err := client.Do(req)
	if err != nil {
		return err
	}

	var data dto.SwitGeneralResponse

	json.NewDecoder(res.Body).Decode(&data)

	if data.Status != "Success" {
		return errors.New("Error canceling: " + res.Status)
	}

	return nil
}

func (ss *switService) ReserveFunds(opts ReserveFundsOpts) (string, error) {
	client := &http.Client{
		Timeout: time.Second * timeout,
	}

	if opts.Description == "" {
		opts.Description = "Automatic charge"
	}

	payloadData, _ := json.Marshal(opts)

	payloadRequest := bytes.NewReader(payloadData)

	req, err := http.NewRequest("POST", ss.cfg.SwitBaseUrl+"/payments", payloadRequest)
	if err != nil {
		return "", err
	}

	ss.addHeaders(req)

	response, err := client.Do(req)
	if err != nil {
		return "", err
	}

	var data dto.SwitGeneralResponse

	json.NewDecoder(response.Body).Decode(&data)

	if data.Status != "Success" {
		return "", ErrProccesingPayment
	}

	result, _ := data.Result.(map[string]any)

	transId := result["transactionId"].(string)

	return transId, nil
}

func (ss *switService) CreateCustomer(cus *schemas.Customer) (string, error) {
	client := &http.Client{
		Timeout: time.Second * timeout,
	}

	payload := struct {
		Email     string `json:"email"`
		FirstName string `json:"first_name"`
		LastName  string `json:"last_name"`
	}{
		Email:     cus.Email,
		FirstName: cus.FirstName,
		LastName:  cus.FirstLastName + " " + cus.SecondLastName,
	}

	copier.Copy(&payload, cus)

	payloadData, _ := json.Marshal(payload)

	payloadRequest := bytes.NewReader(payloadData)

	req, err := http.NewRequest("POST", ss.cfg.SwitBaseUrl+"/customers", payloadRequest)
	if err != nil {
		return "", err
	}

	ss.addHeaders(req)

	response, err := client.Do(req)
	if err != nil {
		return "", err
	}

	if response.StatusCode != 200 {
		return "", errors.New("Error creating customer in swit - " + response.Status)
	}

	var data dto.SwitGeneralResponse

	json.NewDecoder(response.Body).Decode(&data)

	if data.Status != "Success" {
		return "", errors.New("Error creating new customer")
	}

	id, _ := data.Result.(string)

	return id, nil
}

func (ss *switService) ListCardsByCustomer(cusID string) ([]schemas.SwitSource, error) {
	client := &http.Client{
		Timeout: time.Second * timeout,
	}

	req, err := http.NewRequest("GET", ss.cfg.SwitBaseUrl+"/customers/"+cusID+"/cards", nil)
	if err != nil {
		return nil, err
	}

	ss.addHeaders(req)

	res, err := client.Do(req)
	if err != nil {
		return nil, err
	}

	if res.StatusCode != 200 {
		return nil, errors.New("Eror listing customer cards - " + res.Status)
	}

	sources := make([]schemas.SwitSource, 0)

	var data dto.SwitGeneralResponse

	json.NewDecoder(res.Body).Decode(&data)

	if data.Status != "Success" {
		return nil, errors.New("Error listing customer cards")
	}

	for _, s := range data.Result.([]any) {
		v := utils.Transform[schemas.SwitSource](s)

		sources = append(sources, *v)
	}

	return sources, nil
}
